### DISCUSSION TITLE
.NET Core request - make everything mockable, please

### DISCUSSION THREAD

**Discussion by pelletburner** [dotnet/csharplang#6712] [General]
I'm not 100% sure if this is the right place to ask for additions to .NET Core, but one thing that really bothers me for many years now is that most classes in .NET are not mockable.

My request has two aspects:

- Please add interfaces for all classes in the .NET framework
- Please add replacements for all static stuff in the .NET framework which can be mocked

This would really take .NET Core to the next level.

**Comment by En3Tho** [comment]
To the next level... of mockery of programming.

**Reply by pelletburner** [reply]
Whats wrong about having the .NET framework mockable in your opinion? This would help many people a lot.

**Reply by En3Tho** [reply]
Because this adds an extra layer of abstraction where it's not needed at all. Also, what do you mean by framework here? All of it? Should we be able to mock int for example?

If you have concrete problems or use cases then it's fine and please share. Otherwise this has no sense, honestly.

**Reply by pelletburner** [reply]
Things like "DateTime". Thats a classical example, currently most people I talked to about it roll their own wrapper to access ".UtcNow" and things like that in a mockable way.

I think that the framework should support it. Of course not basic types like "int", but everything that operates with sideeffects should be mockable in my opinion.

Yes, "File" and "Directory", too.

**Reply by Joe4evr** [reply]
If you need `DateTime{Offset}` for anything more than taking a timestamp, you should be using NodaTime, which is completely testable and mockable, and battle-hardened against cases that any other "roll your own" implementation likely would never have considered (because dates and times are hard).

For `File` and `Directory` there exists `System.IO.Abstractions` for the same purpose.

These are some of the known friction points since they were designed with a different mentality than today. There's maybe a couple more in this same bucket (and there likely exist suitable abstraction packages for those as well), but there's no reason that *everything* in the BCL needs to be mockable, because it's not *your* responsibility to test the framework.

**Reply by En3Tho** [reply]
https://github.com/dotnet/runtime/issues/36617 - has recent answers from runtime team

**Comment by CyrusNajmabadi** [comment]
Def not a csharplang request.  Consider dotnet/runtime?

**Reply by pelletburner** [reply]
Ok, will do that!

**Comment by qrli** [comment]
There are already plenty of products doing it without the extra needed work you listed, including MS Fakes which comes with Visual Studio.

But in my experience, they are only necessary when the code is not properly designed thus hard to test.
